File: 프로그래머스/브루트포스/타겟 넘버.py
answer = 0


def back_tracking(cnt, n, numbers, target):
    global answer

    if cnt == len(numbers):
        if n == target:
            answer += 1
        return
    
    back_tracking(cnt + 1, n + numbers[cnt], numbers, target)
    back_tracking(cnt + 1, n - numbers[cnt], numbers, target)

        
def solution(numbers, target):
    back_tracking(0, 0, numbers, target)
    return answer

File: 프로그래머스/그래프탐색/타겟 넘버.java
import java.util.*;

class Solution {
    int answer = 0;
    
    public int solution(int[] numbers, int target) {
        dfs(0, 0, 0, numbers, target, numbers.length);
        return answer;
    }
    
    private void dfs(int cur, int start, int cnt, int[] numbers, int target, int n) {
        if (cnt == n) {
            if (cur == target) {
                answer++;
            }
            return;
        }
        
        for (int i = start; i < n; i++) {
            dfs(cur + numbers[i], i + 1, cnt + 1, numbers, target, n);
            dfs(cur - numbers[i], i + 1, cnt + 1, numbers, target, n);
        }
    }
}